### PR TITLE
LocalGrainDirectory: use membership snapshots

### PR DESCRIPTION
**Pull request by ReubenBond**

### CODE REVIEW

File: src/Orleans.Runtime/Catalog/Catalog.cs [hunk 1]
```diff
@@ -1443,5 +1444,15 @@ private void OnSiloStatusChange(SiloAddress updatedSilo, SiloStatus status)
                 }
             }
         }
+
+        private SiloAddress TryGetGrainDirectoryPartitionForDiagnostics(GrainId grainId)
```
Comment (ashtonkj) [inline src/Orleans.Runtime/Catalog/Catalog.cs]: Nit: Isn't it more common for a `TryGetXX` to return a bool value and have an out parameter?

Comment (ReubenBond) [reply src/Orleans.Runtime/Catalog/Catalog.cs]: Certainly is. I'll change it.

File: src/Orleans.Runtime/GrainDirectory/GrainDirectoryHandoffManager.cs [hunk 2]
```diff
@@ -17,8 +17,9 @@ internal class GrainDirectoryHandoffManager
         private const int HANDOFF_CHUNK_SIZE = 500;
         private static readonly TimeSpan RetryDelay = TimeSpan.FromMilliseconds(250);
         private const int MAX_OPERATION_DEQUEUE = 2;
+        private readonly ILocalSiloDetails localSiloDetails;
         private readonly LocalGrainDirectory localDirectory;
-        private readonly ISiloStatusOracle siloStatusOracle;
+        private readonly IClusterMembershipService clusterMembershipService;
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/GrainDirectory/GrainDirectoryHandoffManager.cs]: I would have used `IMembershipService`, but it was taken.

I will rename it

File: src/Orleans.Runtime/GrainDirectory/DirectoryMembershipSnapshot.cs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+using Microsoft.Extensions.Logging;
+using Orleans.Configuration;
+using Orleans.Runtime.Utilities;
+
+namespace Orleans.Runtime.GrainDirectory
+{
+    /// <summary>
+    /// A snapshot of cluster membership from the perspective of the local grain directory.
+    /// </summary>
+    internal class DirectoryMembershipSnapshot
```
[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/GrainDirectory/DirectoryMembershipSnapshot.cs]: They could be extension methods ¯\\\_(ツ)\_/¯

[the remark this answers is not in the document]
Comment (ReubenBond) [reply src/Orleans.Runtime/GrainDirectory/DirectoryMembershipSnapshot.cs]: I hid the data members instead. Eventually we could make this an interface perhaps, but I think it's not generalizable yet.

**Comment by ReubenBond** [comment]
The failure is a known-flaky test (AccountWithLog) - we should skip it

**Pull request merged**